data/hlint.1
.TH HLINT "1" "July 2009" "HLint (C) Neil Mitchell 2006-2009" "User Commands"
.SH NAME
HLint \- haskell source code suggestions
.SH SYNOPSIS
.B hlint
[\fIfiles/directories\fR] [\fIoptions\fR]

.SH DESCRIPTION
\fIHLint\fR is a tool for suggesting possible improvements to Haskell code. These suggestions include ideas such as using alternative functions, simplifying code and spotting redundancies.

.SH OPTIONS
.TP
\fB\-?\fR \fB\-\-help\fR
Display help message
.TP
\fB\-v\fR \fB\-\-version\fR
Display version information
.TP
\fB\-r[file]\fR \fB\-\-report\fR[=\fIfile\fR]
Generate a report in HTML
.TP
\fB\-h\fR \fIfile\fR \fB\-\-hint\fR=\fIfile\fR
Hint/ignore file to use
.TP
\fB\-c\fR \fB\-\-color\fR, \fB\-\-colour\fR
Color the output (requires ANSI terminal)
.TP
\fB\-i\fR \fImessage\fR \fB\-\-ignore\fR=\fImessage\fR
Ignore a particular hint
.TP
\fB\-s\fR \fB\-\-show\fR
Show all ignored ideas
.TP
\fB\-t\fR \fB\-\-test\fR
Run in test mode
.SH EXAMPLE
"To check all Haskell files in 'src' and generate a report type:"
.IP
hlint src \fB\-\-report\fR
.SH "SEE ALSO"
The full documentation for
.B HLint
is available in \fI/usr/share/doc/hlint/hlint.html\fI.
.SH AUTHOR
This   manual   page   was  written  by  Joachim Breitner <nomeata@debian.org>
for the Debian system (but may be used by others).


CHANGES.txt
Changelog for HLint (* = breaking change)

    #651, warn on unused NamedFieldPuns
    #646, switch to an HTML doctype
2.1.22, released 2019-05-25
    #634, suggest modifyIORef ==> writeIORef when applicable
    #642, suggest null in more places
    #640, reenable GHC parsing
2.1.21, released 2019-05-19
    #637, temporary workaround for GHC parser segfaults
2.1.20, released 2019-05-15
*   Fix a dumb break in the API on parseModuleEx
2.1.19, released 2019-05-14
*   Revert PVP breakage
2.1.18, released 2019-05-13
*   Change parseModuleEx/ParseError by accident
    #633, don't suggest changes inside RULES
    #631, suggest typeOf ==> typeRep
    Add matching on type variables
    #627, restrict to GHC 8.4 and above
2.1.17, released 2019-04-17
    #626, add operator wildcards with ?, ??, ??? etc
    #625, fix an rnf/rhs typo
    #562, make test --verbose show a list of matched hints
2.1.16, released 2019-04-15
    Make `seq` and `rem` hints apply to prefix functions
    #604, suggest rnf x `seq` () ==> rnf x
    #619, require haskell-src-exts-util-0.2.5
    #619, fix move guards forward with record puns
    #618, add pure x <*> y ==> x <$> y
    #611, add == and subst for more advanced match conditions
    #612, add: Suggest f =<< instead of maybe Nothing f
    #609, add code smells
    #614, adds refactorings for camelCase and some list suggestions
    #605, make command line arguments override the .yaml file
    #603, QuasiQuotes can programatically use any extension
2.1.15, released 2019-02-27
    #593, reorder guards in list comps where possible
    #597, suggest pushing a map over a list comp inside
    Say redundant pure, when the thing you are removing is pure
    #554, add more verbosity
    Don't test with GHC 7.4 to 7.8
    #590, say which extensions should be deleted
    #591, be careful about encoding on stdin
2.1.14, released 2019-01-28
    #587, fix extensions implied by ImplicitParams
    #588, suggest optional from attoparsec
2.1.13, released 2019-01-23
    #583, suggest left sections to avoid lambdas
    #580, remove redundant LANGUAGE pragmas which are implied by others
    #575, add fixities for lattice
    #564, fix hint around withFile with AppendMode
2.1.12, released 2018-12-10
    Require haskell-src-exts-1.21
2.1.11, released 2018-12-02
    #553, define __HLINT__=1 for the C preprocessor
    #546, suggest `x $> y` for `const x <$> y`, `pure x <$> y`, and `return x <$> y`
    #546, suggest `x <$ y` for `x <&> const y`, `x <&> pure y`, and `x <&> return y`
    #556, disable a few incorrect lens hints
    #545, don't suggest turning type applications into sections
    #466, avoid false positives for Esqueleto
    #535, more lens hints
    Allow {-# HLINT #-} and {- HLINT -} pragmas
    #532, generate requested report files even if there are no hints
    #524, don't suggest newtype for existentials
    #521, add a hint for f x@_ = ... ==> f x = ...
2.1.10, released 2018-08-16
    #516, don't require a .hlint.yaml when running tests
    Prefer .hlint.yaml to HLint.hs for settings
    #513, add section links in the HTML report
2.1.9, released 2018-08-08
    Add QuickCheck fixities
    Warn on redundant EmptyCase extension
2.1.8, released 2018-07-06
    #509, remove incorrect suggestions around sequence/pure
2.1.7, released 2018-07-03
    #483, don't break quasi quotes when suggesting const
    #404, remove the "Unnecessary hiding" hint introduced in #338
    #162, make avoiding lambda with `infix` give a different name
    #507, rename id x ==> x to redundant id
    #286, improve the duplicate pragma message
    #399, suggest (& f) ==> f
    #136, don't suggest eta-reducing runST
    #345, add catMaybes/fmap ==> mapMaybe
    #345, add foldMap id ==> fold
    #364, suggest >> instead of >>= \_ ->
    #502, DeriveTraversable implies DeriveFoldable and DeriveFunctor
    Add hints about fusing traverse/map
    Better names for mapM/map fusion hints and others
    #498, change the output to say "Perhaps:" rather than "Why not:"
2.1.6, released 2018-06-16
    Match on explicit brackets at the root of a match expression
    #470, suggest TupleSections
    #496, suggest sequence/fmap ==> mapM
    #473, warn on redundant void, _ <- and return ()
    Make use of <$> more general, but in simpler cases
    Warn about returns in the middle of do blocks
    #471, suggest readTVarIO
    #468, suggest using sortOn/Down
    #458, document the restriction feature
    #494, don't suggest newtype for unboxed tuples
    #488, avoid warning about more test prefixes
2.1.5, released 2018-05-05
    #478, take account of deriving strategies for extension use
    #477, don't warn about unit_ as tasty-discover recommends it
2.1.4, released 2018-05-01
    Don't warn about redundant $ for a $ b{c=d}
2.1.3, released 2018-04-18
    Improve the performance of the camelCase hint
    Don't suggest camelCase for record fields
    Add a --timing flag to detect what is slow
2.1.2, released 2018-04-16
    #407, don't error on unknown extensions on the command line
    Require extra-1.6.6
    #464, add more hints for concatMap
    #462, ignore home directory when it isn't present
2.1.1, released 2018-03-24
    #457, suggest turning on LambdaCase if necessary
    #457, add RequiresExtension note
    #454, add fixities for the HSpec `should*` functions
    #455, add some more sequence hints
    #453, allow pure in a few Monad hints as well as return
    #451, add --with-group command line option
    #424, suggest Foldable.forM_ in a few more places
    #445, add suggestions for reader/state monad
    #443, suggest join (x <$> y) ==> x =<< y
2.1, released 2018-02-07
*   #433, make ideas span multiple modules/declarations
    #433, allow ignoring statement-level duplication hint
    #439, add more fixities for new base operators
    #437, --json output should be finite
    #425, avoid misparsing use of Gtk2Hs `on` function
    #353, detect unused results from for/traverse/sequenceA

src/HLint.hs
{-# LANGUAGE RecordWildCards, TupleSections #-}
{-# OPTIONS_GHC -fno-warn-incomplete-patterns #-}

module HLint(hlint, readAllSettings) where

import Control.Applicative
import Control.Monad.Extra
import Control.Exception
import Control.Concurrent.Extra
import System.Console.CmdArgs.Verbosity
import Data.List.Extra
import GHC.Conc
import System.Exit
import System.IO.Extra
import System.Time.Extra
import Data.Tuple.Extra
import Prelude

import Data.Version.Extra
import System.Process.Extra
import Data.Maybe
import System.Directory
import System.FilePath

import CmdLine
import Config.Read
import Config.Type
import Config.Compute
import Report
import Idea
import Apply
import Test.All
import Hint.All
import Grep
import Timing
import Test.Proof
import Parallel
import HSE.All
import CC


-- | This function takes a list of command line arguments, and returns the given hints.
--   To see a list of arguments type @hlint --help@ at the console.
--   This function writes to the stdout/stderr streams, unless @--quiet@ is specified.
--
--   As an example:
--
-- > do hints <- hlint ["src", "--ignore=Use map","--quiet"]
-- >    when (length hints > 3) $ error "Too many hints!"
--
--   /Warning:/ The flags provided by HLint are relatively stable, but do not have the same
--   API stability guarantees as the rest of the strongly-typed API. Do not run this function
--   on a your server with untrusted input.
hlint :: [String] -> IO [Idea]
hlint args = do
    cmd <- getCmd args
    case cmd of
        CmdMain{} -> do
            startTimings
            (time, xs) <- duration $ hlintMain args cmd
            when (cmdTiming cmd) $ do
                printTimings
                putStrLn $ "Took " ++ showDuration time
            return $ if cmdNoExitCode cmd then [] else xs
        CmdGrep{} -> hlintGrep cmd >> return []
        CmdHSE{}  -> hlintHSE  cmd >> return []
        CmdTest{} -> hlintTest cmd >> return []

hlintHSE :: Cmd -> IO ()
hlintHSE c@CmdHSE{..} = do
    v <- getVerbosity
    forM_ cmdFiles $ \x -> do
        putStrLn $ "Parse result of " ++ x ++ ":"
        let (lang,exts) = cmdExtensions c
        -- We deliberately don't use HSE.All here to avoid any bugs in HLint
        -- polluting our bug reports (which is the main use of HSE)
        res <- parseFileWithMode defaultParseMode{baseLanguage=lang, extensions=exts} x
        case res of
            x@ParseFailed{} -> print x
            ParseOk m -> case v of
                Loud -> print m
                Quiet -> print $ prettyPrint m
                _ -> print $ void m
        putStrLn ""

hlintTest :: Cmd -> IO ()
hlintTest cmd@CmdTest{..} =
    if not $ null cmdProof then do
        files <- cmdHintFiles cmd
        s <- readFilesConfig $ map (,Nothing) files
        let reps = if cmdReports == ["report.html"] then ["report.txt"] else cmdReports
        mapM_ (proof reps s) cmdProof
     else do
        failed <- test cmd (\args -> do errs <- hlint args; unless (null errs) $ exitWith $ ExitFailure 1) cmdDataDir cmdGivenHints
        when (failed > 0) exitFailure

cmdParseFlags :: Cmd -> ParseFlags
cmdParseFlags cmd = parseFlagsSetLanguage (cmdExtensions cmd) $ defaultParseFlags{cppFlags=cmdCpp cmd}

hlintGrep :: Cmd -> IO ()
hlintGrep cmd@CmdGrep{..} =
    if null cmdFiles then
        exitWithHelp
     else do
        files <- concatMapM (resolveFile cmd Nothing) cmdFiles
        if null files then
            error "No files found"
         else
            runGrep cmdPattern (cmdParseFlags cmd) files

withVerbosity :: Verbosity -> IO a -> IO a
withVerbosity new act = do
    old <- getVerbosity
    (setVerbosity new >> act) `finally` setVerbosity old

hlintMain :: [String] -> Cmd -> IO [Idea]
hlintMain args cmd@CmdMain{..}
    | cmdDefault = do
        ideas <- if null cmdFiles then return [] else withVerbosity Quiet $
            runHlintMain args cmd{cmdJson=False,cmdSerialise=False,cmdRefactor=False} Nothing
        let bad = nubOrd $ map ideaHint ideas
        src <- readFile $ cmdDataDir </> "default.yaml"
        if null bad then putStr src else do
            let group1:groups = splitOn ["",""] $ lines src
            let group2 = "# Warnings currently triggered by your code" :
                         ["- ignore: {name: " ++ show x ++ "}" | x <- bad]
            putStr $ unlines $ intercalate ["",""] $ group1:group2:groups
        return []
    | null cmdFiles && not (null cmdFindHints) = do
        hints <- concatMapM (resolveFile cmd Nothing) cmdFindHints
        mapM_ (putStrLn . fst <=< computeSettings (cmdParseFlags cmd)) hints >> return []
    | null cmdFiles =
        exitWithHelp
    | cmdRefactor =
        withTempFile $ runHlintMain args cmd . Just
    | otherwise =
        runHlintMain args cmd Nothing

runHlintMain :: [String] -> Cmd -> Maybe FilePath -> IO [Idea]
runHlintMain args cmd tmpFile = do
    (cmd, settings) <- readAllSettings args cmd
    runHints args settings =<< resolveFiles cmd tmpFile

resolveFiles :: Cmd -> Maybe FilePath -> IO Cmd
resolveFiles cmd@CmdMain{..} tmpFile = do
    files <- concatMapM (resolveFile cmd tmpFile) cmdFiles
    if null files
        then error "No files found"
        else pure cmd { cmdFiles = files }
resolveFiles cmd _ = pure cmd

readAllSettings :: [String] -> Cmd -> IO (Cmd, [Setting])
readAllSettings args1 cmd@CmdMain{..} = do
    files <- cmdHintFiles cmd
    settings1 <-
        readFilesConfig $
        map (,Nothing) files
        ++ [("CommandLine.hs",Just x) | x <- cmdWithHints]
        ++ [("CommandLine.yaml",Just (enableGroup x)) | x <- cmdWithGroups]
    let args2 = [x | SettingArgument x <- settings1]
    cmd@CmdMain{..} <- if null args2 then return cmd else getCmd $ args2 ++ args1 -- command line arguments are passed last
    settings2 <- concatMapM (fmap snd . computeSettings (cmdParseFlags cmd)) cmdFindHints
    settings3 <- return [SettingClassify $ Classify Ignore x "" "" | x <- cmdIgnore]
    return (cmd, settings1 ++ settings2 ++ settings3)
    where
        enableGroup groupName =
            unlines
            ["- group:"
            ,"    name: " ++ groupName
            ,"    enabled: true"
            ]

runHints :: [String] -> [Setting] -> Cmd -> IO [Idea]
runHints args settings cmd@CmdMain{..} = do
    j <- if cmdThreads == 0 then getNumProcessors else return cmdThreads
    withNumCapabilities j $ do
        let outStrLn = whenNormal . putStrLn
        ideas <- getIdeas cmd settings
        ideas <- return $ if cmdShowAll then ideas else  filter (\i -> ideaSeverity i /= Ignore) ideas
        if cmdJson then
            putStrLn $ showIdeasJson ideas
         else if cmdCC then
            mapM_ (printIssue . fromIdea) ideas
         else if cmdSerialise then do
            hSetBuffering stdout NoBuffering
            print $ map (show &&& ideaRefactoring) ideas
         else if cmdRefactor then
            handleRefactoring ideas cmdFiles cmd
         else do
            usecolour <- cmdUseColour cmd
            showItem <- if usecolour then showANSI else return show
            mapM_ (outStrLn . showItem) ideas
            handleReporting ideas cmd
        return ideas

getIdeas :: Cmd -> [Setting] -> IO [Idea]
getIdeas cmd@CmdMain{..} settings = do
    settings <- return $ settings ++ map (Builtin . fst) builtinHints
    let flags = cmdParseFlags cmd
    ideas <- if cmdCross
        then applyHintFiles flags settings cmdFiles
        else concat <$> parallel cmdThreads [evaluateList =<< applyHintFile flags settings x Nothing | x <- cmdFiles]
    return $ if not (null cmdOnly)
        then [i | i <- ideas, ideaHint i `elem` cmdOnly]
        else ideas

handleRefactoring :: [Idea] -> [String] -> Cmd -> IO ()
handleRefactoring ideas files cmd@CmdMain{..} =
    case cmdFiles of
        [file] -> do
            -- Ensure that we can find the executable
            path <- checkRefactor (if cmdWithRefactor == "" then Nothing else Just cmdWithRefactor)
            -- writeFile "hlint.refact"
            let hints =  show $ map (show &&& ideaRefactoring) ideas
            withTempFile $ \f -> do
                writeFile f hints
                exitWith =<< runRefactoring path file f cmdRefactorOptions
        _ -> error "Refactor flag can only be used with an individual file"


handleReporting :: [Idea] -> Cmd -> IO ()
handleReporting showideas cmd@CmdMain{..} = do
    let outStrLn = whenNormal . putStrLn
    forM_ cmdReports $ \x -> do
        outStrLn $ "Writing report to " ++ x ++ " ..."
        writeReport cmdDataDir x showideas
    unless cmdNoSummary $ do
        let n = length showideas
        outStrLn $ if n == 0 then "No hints" else show n ++ " hint" ++ ['s' | n/=1]

runRefactoring :: FilePath -> FilePath -> FilePath -> String -> IO ExitCode
runRefactoring rpath fin hints opts =  do
    let args = [fin, "-v0"] ++ words opts ++ ["--refact-file", hints]
    (_, _, _, phand) <- createProcess $ proc rpath args
    try $ hSetBuffering stdin LineBuffering :: IO (Either IOException ())
    hSetBuffering stdout LineBuffering
    -- Propagate the exit code from the spawn process
    waitForProcess phand

checkRefactor :: Maybe FilePath -> IO FilePath
checkRefactor rpath = do
    let excPath = fromMaybe "refactor" rpath
    mexc <- findExecutable excPath
    case mexc of
        Just exc ->  do
            ver <- readVersion . tail <$> readProcess exc ["--version"] ""
            if versionBranch ver >= [0,1,0,0]
                then return exc
                else error "Your version of refactor is too old, please upgrade to the latest version"
        Nothing -> error $ unlines [ "Could not find refactor", "Tried with: " ++ excPath ]

evaluateList :: [a] -> IO [a]
evaluateList xs = do
    evaluate $ length xs
    return xs
